orchagent/portsorch.cpp

```diff
@@ -638,6 +567,312 @@ PortsOrch::PortsOrch(DBConnector *db, DBConnector *stateDb, vector<table_name_wi
+
+    status = sai_switch_api->get_switch_attribute(gSwitchId, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get port number, rv:%d", status);
+        auto handle_status = handleSaiGetStatus(SAI_API_SWITCH, status);
+        if (handle_status != task_process_status::task_success)
+        {
+            SWSS_LOG_THROW("PortsOrch initialization failure");
+        }
+    }
+
+    this->m_portCount = attr.value.u32;
+
+    SWSS_LOG_NOTICE("Get %d ports", this->m_portCount);
+
+    // Get port list
+    std::vector<sai_object_id_t> portList(this->m_portCount, SAI_NULL_OBJECT_ID);
+
+    attr.id = SAI_SWITCH_ATTR_PORT_LIST;
+    attr.value.objlist.count = static_cast<sai_uint32_t>(portList.size());
+    attr.value.objlist.list = portList.data();
+
+    status = sai_switch_api->get_switch_attribute(gSwitchId, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get port list, rv:%d", status);
+        auto handle_status = handleSaiGetStatus(SAI_API_SWITCH, status);
+        if (handle_status != task_process_status::task_success)
+        {
+            SWSS_LOG_THROW("PortsOrch initialization failure");
+        }
+    }
+
+    // Get port hardware lane info
+    for (const auto &portId : portList)
+    {
+        std::vector<sai_uint32_t> laneList(Port::max_lanes, 0);
+
+        attr.id = SAI_PORT_ATTR_HW_LANE_LIST;
+        attr.value.u32list.count = static_cast<sai_uint32_t>(laneList.size());
+        attr.value.u32list.list = laneList.data();
+
+        status = sai_port_api->get_port_attribute(portId, 1, &attr);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to get hardware lane list pid:%" PRIx64, portId);
+            auto handle_status = handleSaiGetStatus(SAI_API_PORT, status);
+            if (handle_status != task_process_status::task_success)
+            {
+                SWSS_LOG_THROW("PortsOrch initialization failure");
+            }
+        }
+
+        std::set<std::uint32_t> laneSet;
+        for (sai_uint32_t i = 0; i < attr.value.u32list.count; i++)
+        {
+            laneSet.insert(attr.value.u32list.list[i]);
+        }
+
+        this->m_portListLaneMap[laneSet] = portId;
+
+        SWSS_LOG_NOTICE(
+            "Get port with lanes pid:%" PRIx64 " lanes:%s",
+            portId, swss::join(" ", laneSet.cbegin(), laneSet.cend()).c_str()
+        );
+    }
+}
+
+auto PortsOrch::getPortConfigState() const -> port_config_state_t
+{
+    return this->m_portConfigState;
+}
+
+void PortsOrch::setPortConfigState(port_config_state_t value)
+{
+    this->m_portConfigState = value;
+}
+
+bool PortsOrch::addPortBulk(const std::vector<PortConfig> &portList)
+{
+    // The method is used to create ports in a bulk mode.
+    // The action takes place when:
+    // 1. Ports are being initialized at system start
+    // 2. Ports are being added/removed by a user at runtime
```

Comment:
@nazariig #2 is done via breakout CLI? What other way user can use?

Reply:
@prgeor this means that api will be invoked each time there is an attempt to modify port related stuff in db. Breakout CLI does exactly the same